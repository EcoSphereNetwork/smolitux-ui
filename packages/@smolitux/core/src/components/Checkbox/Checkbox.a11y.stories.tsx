import React from 'react';
import type { Meta, StoryObj } from '@storybook/react';
import { Checkbox.a11y } from './Checkbox.a11y';

const meta: Meta<typeof Checkbox.a11y> = {
  title: 'Components/Checkbox.a11y',
  component: Checkbox.a11y,
  parameters: {
    layout: 'centered',
  },
  tags: ['autodocs'],
};

export default meta;
type Story = StoryObj<typeof meta>;

export const Default: Story = {
  args: {
    children: 'Checkbox.a11y',
  },
};

export const CustomStyle: Story = {
  args: {
    children: 'Custom Checkbox.a11y',
    className: 'custom-style',
  },
};

export const Interactive: Story = {
  args: {
    children: 'Interactive Checkbox.a11y',
    onClick: () => alert('Clicked!'),
  },
};
