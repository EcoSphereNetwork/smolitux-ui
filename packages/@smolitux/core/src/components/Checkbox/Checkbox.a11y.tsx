import React, { forwardRef, useRef, useEffect, useState, useId } from 'react';
import { useFormControl } from '../FormControl';

export type CheckboxSize = 'xs' | 'sm' | 'md' | 'lg' | 'xl';
export type CheckboxVariant = 'solid' | 'outline' | 'filled' | 'minimal';
export type CheckboxColorScheme = 'primary' | 'secondary' | 'success' | 'danger' | 'warning' | 'info' | 'neutral';
export type CheckboxLabelPosition = 'left' | 'right';
export type CheckboxDisplayType = 'checkbox' | 'switch' | 'radio' | 'toggle' | 'button';

export interface CheckboxProps extends Omit<React.InputHTMLAttributes<HTMLInputElement>, 'type' | 'size'> {
  /** Text-Label */
  label?: React.ReactNode;
  /** Hilfetext */
  helperText?: React.ReactNode;
  /** Fehlermeldung */
  error?: React.ReactNode;
  /** Erfolgsmeldung */
  successMessage?: React.ReactNode;
  /** Größe der Checkbox */
  size?: CheckboxSize;
  /** Visuelle Variante */
  variant?: CheckboxVariant;
  /** Indeterminierter Zustand */
  indeterminate?: boolean;
  /** Farbe der Checkbox */
  colorScheme?: CheckboxColorScheme;
  /** Ob die Checkbox einen Rahmen haben soll */
  bordered?: boolean;
  /** Ob die Checkbox abgerundete Ecken haben soll */
  rounded?: boolean;
  /** Ob die Checkbox einen Schatten haben soll */
  shadow?: boolean;
  /** Ob die Checkbox einen Hover-Effekt haben soll */
  hoverable?: boolean;
  /** Ob die Checkbox einen Fokus-Effekt haben soll */
  focusable?: boolean;
  /** Ob die Checkbox einen Übergangseffekt haben soll */
  transition?: boolean;
  /** Ob die Checkbox einen transparenten Hintergrund haben soll */
  transparent?: boolean;
  /** Ob die Checkbox einen Tooltip haben soll */
  tooltip?: string;
  /** Ob die Checkbox im Ladezustand ist */
  isLoading?: boolean;
  /** Ob die Checkbox gültig ist */
  isValid?: boolean;
  /** Ob die Checkbox ungültig ist */
  isInvalid?: boolean;
  /** Ob die Checkbox erfolgreich validiert ist */
  isSuccess?: boolean;
  /** Ob die Checkbox deaktiviert ist */
  isDisabled?: boolean;
  /** Ob die Checkbox erforderlich ist */
  isRequired?: boolean;
  /** Ob die Checkbox schreibgeschützt ist */
  isReadOnly?: boolean;
  /** Ob die Checkbox als Switch dargestellt werden soll */
  isSwitch?: boolean;
  /** Ob die Checkbox als Toggle dargestellt werden soll */
  isToggle?: boolean;
  /** Ob die Checkbox als Button dargestellt werden soll */
  isButton?: boolean;
  /** Position des Labels */
  labelPosition?: CheckboxLabelPosition;
  /** Ob das Label versteckt werden soll (nur für Screenreader sichtbar) */
  hideLabel?: boolean;
  /** Zusätzliche CSS-Klassen für das Label */
  labelClassName?: string;
  /** Zusätzliche CSS-Klassen für den Hilfetext */
  helperTextClassName?: string;
  /** Zusätzliche CSS-Klassen für die Fehlermeldung */
  errorClassName?: string;
  /** Zusätzliche CSS-Klassen für die Erfolgsmeldung */
  successClassName?: string;
  /** Zusätzliche CSS-Klassen für den Container */
  containerClassName?: string;
  /** Zusätzliche CSS-Klassen für die Checkbox */
  checkboxClassName?: string;
  /** Zusätzliche CSS-Klassen für den Wrapper */
  wrapperClassName?: string;
  /** ARIA-Label für die Checkbox */
  ariaLabel?: string;
  /** Beschreibung für die Checkbox (für Screenreader) */
  description?: string;
  /** Ob die Checkbox als "busy" markiert werden soll */
  busy?: boolean;
  /** Ob die Checkbox als "polite" oder "assertive" angekündigt werden soll */
  liveRegionPoliteness?: 'polite' | 'assertive' | 'off';
  /** Ob die Checkbox als "atomic" angekündigt werden soll */
  atomic?: boolean;
  /** Ob die Checkbox als "relevant" angekündigt werden soll */
  relevant?: 'additions' | 'removals' | 'text' | 'all';
  /** Testid für automatisierte Tests */
  testId?: string;
}

/**
 * Barrierefreie Checkbox-Komponente für Formulare
 * 
 * @example
 * ```tsx
 * <CheckboxA11y 
 *   label="Ich stimme den AGB zu" 
 *   isRequired
 *   description="Bitte bestätigen Sie, dass Sie unsere Allgemeinen Geschäftsbedingungen gelesen haben und ihnen zustimmen."
 * />
 * 
 * <CheckboxA11y 
 *   label="Newsletter abonnieren" 
 *   helperText="Sie können sich jederzeit wieder abmelden."
 *   isSwitch
 * />
 * ```
 */
export const CheckboxA11y = forwardRef<HTMLInputElement, CheckboxProps>(({
  label,
  helperText,
  error,
  successMessage,
  size = 'md',
  variant = 'solid',
  indeterminate = false,
  colorScheme = 'primary',
  bordered = false,
  rounded = true,
  shadow = false,
  hoverable = true,
  focusable = true,
  transition = true,
  transparent = false,
  tooltip,
  isLoading = false,
  isValid = false,
  isInvalid = false,
  isSuccess = false,
  isDisabled = false,
  isRequired = false,
  isReadOnly = false,
  isSwitch = false,
  isToggle = false,
  isButton = false,
  labelPosition = 'right',
  hideLabel = false,
  labelClassName = '',
  helperTextClassName = '',
  errorClassName = '',
  successClassName = '',
  containerClassName = '',
  checkboxClassName = '',
  wrapperClassName = '',
  ariaLabel,
  description,
  busy = false,
  liveRegionPoliteness = 'polite',
  atomic = true,
  relevant,
  testId,
  ...props
}, ref) => {
  // Generiere eindeutige ID für die Checkbox
  const uniqueIdFromHook = useId();
  const uniqueId = props.id || `checkbox-${uniqueIdFromHook}`;
  
  // Refs
  const checkboxRef = useRef<HTMLInputElement>(null);
  
  // State
  const [isFocused, setIsFocused] = useState(false);
  const [isChecked, setIsChecked] = useState<boolean>(props.checked !== undefined ? Boolean(props.checked) : props.defaultChecked !== undefined ? Boolean(props.defaultChecked) : false);
  const [announceMessage, setAnnounceMessage] = useState<string>('');
  
  // Hole FormControl-Context, falls vorhanden
  const formControl = useFormControl();
  
  // Kombiniere Props mit FormControl-Context
  const _disabled = isDisabled || formControl?.disabled || false;
  const _required = isRequired || formControl?.required || false;
  const _readOnly = isReadOnly || formControl?.readOnly || false;
  const _isInvalid = isInvalid || Boolean(error) || formControl?.isInvalid || false;
  const _isValid = isValid || formControl?.isValid || false;
  const _isSuccess = isSuccess || formControl?.isSuccess || false;
  const _isLoading = isLoading || formControl?.isLoading || false;
  
  // Kombiniere den externen Ref mit unserem internen Ref
  const handleRef = (element: HTMLInputElement | null) => {
    if (checkboxRef) {
      (checkboxRef as React.MutableRefObject<HTMLInputElement | null>).current = element;
    }
    
    if (typeof ref === 'function') {
      ref(element);
    } else if (ref) {
      (ref as React.MutableRefObject<HTMLInputElement | null>).current = element;
    }
  };
  
  // Effekt für indeterminate-Zustand
  useEffect(() => {
    if (checkboxRef.current) {
      checkboxRef.current.indeterminate = indeterminate;
    }
  }, [indeterminate]);
  
  // Effekt für checked-Zustand
  useEffect(() => {
    if (props.checked !== undefined) {
      setIsChecked(Boolean(props.checked));
    }
  }, [props.checked]);
  
  // Klassen für verschiedene Größen
  const sizeClasses = {
    xs: 'h-3 w-3',
    sm: 'h-4 w-4',
    md: 'h-5 w-5',
    lg: 'h-6 w-6',
    xl: 'h-7 w-7'
  };
  
  // Klassen für verschiedene Varianten
  const variantClasses = {
    solid: 'bg-white dark:bg-gray-800 border-2 border-gray-300 dark:border-gray-600',
    outline: 'bg-transparent border-2 border-gray-300 dark:border-gray-600',
    filled: 'bg-gray-100 dark:bg-gray-700 border-2 border-transparent',
    minimal: 'bg-transparent border-0'
  };
  
  // Klassen für verschiedene Farbschemata
  const colorSchemeClasses = {
    primary: 'checked:bg-primary-500 checked:border-primary-500 dark:checked:bg-primary-400 dark:checked:border-primary-400',
    secondary: 'checked:bg-secondary-500 checked:border-secondary-500 dark:checked:bg-secondary-400 dark:checked:border-secondary-400',
    success: 'checked:bg-green-500 checked:border-green-500 dark:checked:bg-green-400 dark:checked:border-green-400',
    danger: 'checked:bg-red-500 checked:border-red-500 dark:checked:bg-red-400 dark:checked:border-red-400',
    warning: 'checked:bg-yellow-500 checked:border-yellow-500 dark:checked:bg-yellow-400 dark:checked:border-yellow-400',
    info: 'checked:bg-blue-500 checked:border-blue-500 dark:checked:bg-blue-400 dark:checked:border-blue-400',
    neutral: 'checked:bg-gray-500 checked:border-gray-500 dark:checked:bg-gray-400 dark:checked:border-gray-400'
  };
  
  // Effekt-spezifische Klassen
  const effectClasses = {
    rounded: rounded ? 'rounded' : '',
    shadow: shadow ? 'shadow-sm' : '',
    hover: hoverable && !_disabled ? 'hover:border-gray-400 dark:hover:border-gray-500' : '',
    focus: focusable && !_disabled ? 'focus:outline-none focus:ring-2 focus:ring-offset-2 focus:ring-primary-500 dark:focus:ring-primary-400' : '',
    transition: transition ? 'transition-all duration-200' : '',
    error: _isInvalid ? 'border-red-500 dark:border-red-400 focus:ring-red-500 focus:border-red-500 dark:focus:ring-red-400 dark:focus:border-red-400' : '',
    success: _isSuccess ? 'border-green-500 dark:border-green-400 focus:ring-green-500 focus:border-green-500 dark:focus:ring-green-400 dark:focus:border-green-400' : '',
    disabled: _disabled ? 'opacity-50 cursor-not-allowed' : '',
    readOnly: _readOnly ? 'cursor-default' : '',
    transparent: transparent ? 'bg-transparent' : ''
  };
  
  // Kombiniere alle Klassen für die Checkbox
  const checkboxClasses = [
    'form-checkbox appearance-none',
    sizeClasses[size],
    variantClasses[variant],
    colorSchemeClasses[colorScheme],
    effectClasses.rounded,
    effectClasses.shadow,
    effectClasses.hover,
    effectClasses.focus,
    effectClasses.transition,
    effectClasses.error,
    effectClasses.success,
    effectClasses.disabled,
    effectClasses.readOnly,
    effectClasses.transparent,
    checkboxClassName
  ].filter(Boolean).join(' ');
  
  // Klassen für den Container
  const containerClasses = [
    'flex items-center',
    labelPosition === 'left' ? 'flex-row-reverse' : 'flex-row',
    containerClassName
  ].filter(Boolean).join(' ');
  
  // Klassen für das Label
  const labelClasses = [
    'text-sm font-medium text-gray-700 dark:text-gray-300',
    labelPosition === 'left' ? 'mr-2' : 'ml-2',
    _disabled ? 'opacity-50 cursor-not-allowed' : '',
    hideLabel ? 'sr-only' : '',
    labelClassName
  ].filter(Boolean).join(' ');
  
  // Event-Handler
  const handleFocus = (e: React.FocusEvent<HTMLInputElement>) => {
    setIsFocused(true);
    
    if (props.onFocus) {
      props.onFocus(e);
    }
  };
  
  const handleBlur = (e: React.FocusEvent<HTMLInputElement>) => {
    setIsFocused(false);
    
    if (props.onBlur) {
      props.onBlur(e);
    }
  };
  
  const handleChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    // Wenn readonly oder disabled, verhindern wir die Änderung
    if (_readOnly || _disabled) {
      e.preventDefault();
      return;
    }
    
    setIsChecked(e.target.checked);
    
    if (props.onChange) {
      props.onChange(e);
    }
    
    // Ankündige für Screenreader
    setAnnounceMessage(e.target.checked ? 'Ausgewählt' : 'Nicht ausgewählt');
  };
  
  const handleKeyDown = (e: React.KeyboardEvent<HTMLInputElement>) => {
    if (_readOnly && (e.key === ' ' || e.key === 'Enter')) {
      e.preventDefault();
      return;
    }
    
    if (props.onKeyDown) {
      props.onKeyDown(e);
    }
  };
  
  // Rendere die Beschreibung (für Screenreader)
  const renderDescription = () => {
    if (!description) return null;
    
    return (
      <div id={`${uniqueId}-description`} className="sr-only">
        {description}
      </div>
    );
  };
  
  // Rendere die Live-Region für Ankündigungen
  const renderLiveRegion = () => {
    return (
      <div 
        aria-live={liveRegionPoliteness} 
        aria-atomic={atomic} 
        aria-relevant={relevant}
        className="sr-only"
      >
        {announceMessage}
      </div>
    );
  };
  
  // Bestimme die Rolle basierend auf dem Typ
  const role = isSwitch || isToggle ? 'switch' : undefined;
  
  // Bestimme den aria-checked-Wert
  const ariaChecked = indeterminate ? 'mixed' : isChecked;
  
  // Bestimme die aria-describedby-Werte
  const ariaDescribedBy = [
    description ? `${uniqueId}-description` : null,
    helperText ? `${uniqueId}-helper` : null,
    error && _isInvalid ? `${uniqueId}-error` : null,
    successMessage && _isSuccess ? `${uniqueId}-success` : null
  ].filter(Boolean).join(' ') || undefined;
  
  return (
    <div className={`${containerClasses}`} title={tooltip} data-testid={testId || "checkbox-container"}>
      {renderDescription()}
      {renderLiveRegion()}
      
      <input
        ref={handleRef}
        id={uniqueId}
        type="checkbox"
        className={checkboxClasses}
        disabled={_disabled}
        readOnly={_readOnly}
        checked={isChecked}
        aria-checked={ariaChecked}
        aria-invalid={_isInvalid}
        aria-required={_required}
        aria-disabled={_disabled}
        aria-readonly={_readOnly}
        aria-describedby={ariaDescribedBy}
        aria-label={!label ? ariaLabel : undefined}
        aria-labelledby={label ? `${uniqueId}-label` : undefined}
        aria-busy={busy || _isLoading}
        role={role}
        aria-roledescription={isSwitch || isToggle ? 'Schalter' : undefined}
        onFocus={handleFocus}
        onBlur={handleBlur}
        onChange={handleChange}
        onKeyDown={handleKeyDown}
        data-testid="checkbox-input"
        {...props}
      />
      
      {label && (
        <label 
          id={`${uniqueId}-label`}
          htmlFor={uniqueId} 
          className={labelClasses}
        >
          {label}
          {_required && <span className="text-red-500 ml-1" aria-hidden="true">*</span>}
          {_required && <span className="sr-only">(Erforderlich)</span>}
        </label>
      )}
      
      {(_isInvalid || helperText || _isSuccess) && (
        <div className="mt-1 text-sm w-full">
          {_isInvalid && error ? (
            <p 
              id={`${uniqueId}-error`} 
              className={`text-red-600 dark:text-red-400 ${errorClassName}`}
              role="alert"
              aria-live="assertive"
            >
              {error}
            </p>
          ) : _isSuccess && successMessage ? (
            <p 
              id={`${uniqueId}-success`} 
              className={`text-green-600 dark:text-green-400 ${successClassName}`}
              role="status"
              aria-live="polite"
            >
              {successMessage}
            </p>
          ) : helperText ? (
            <p 
              id={`${uniqueId}-helper`} 
              className={`text-gray-500 dark:text-gray-400 ${helperTextClassName}`}
              aria-live="polite"
            >
              {helperText}
            </p>
          ) : null}
        </div>
      )}
    </div>
  );
});

CheckboxA11y.displayName = 'CheckboxA11y';

export default CheckboxA11y;