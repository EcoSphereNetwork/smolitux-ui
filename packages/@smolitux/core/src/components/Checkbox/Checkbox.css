.smolitux-checkbox-container {
  display: inline-flex;
  align-items: flex-start;
  position: relative;
  cursor: pointer;
}

.smolitux-checkbox-container--disabled {
  cursor: not-allowed;
  opacity: 0.7;
}

.smolitux-checkbox-input {
  position: absolute;
  opacity: 0;
  width: 0;
  height: 0;
  margin: 0;
}

.smolitux-checkbox-control {
  position: relative;
  display: inline-flex;
  align-items: center;
  justify-content: center;
  flex-shrink: 0;
  width: 1rem;
  height: 1rem;
  border: 2px solid var(--checkbox-border-color, #d1d5db);
  background-color: var(--checkbox-bg, white);
  transition: all 0.2s ease-in-out;
}

.smolitux-checkbox-control--rounded {
  border-radius: 0.25rem;
}

.smolitux-checkbox-control--shadow {
  box-shadow: 0 1px 2px 0 rgba(0, 0, 0, 0.05);
}

.smolitux-checkbox-input:checked + .smolitux-checkbox-control {
  background-color: var(--checkbox-checked-bg, #3b82f6);
  border-color: var(--checkbox-checked-border-color, #3b82f6);
}

.smolitux-checkbox-input:indeterminate + .smolitux-checkbox-control {
  background-color: var(--checkbox-indeterminate-bg, #3b82f6);
  border-color: var(--checkbox-indeterminate-border-color, #3b82f6);
}

.smolitux-checkbox-input:focus + .smolitux-checkbox-control {
  outline: 2px solid transparent;
  outline-offset: 2px;
  box-shadow: 0 0 0 3px rgba(59, 130, 246, 0.2);
}

.smolitux-checkbox-input:disabled + .smolitux-checkbox-control {
  background-color: var(--checkbox-disabled-bg, #f3f4f6);
  border-color: var(--checkbox-disabled-border-color, #d1d5db);
}

.smolitux-checkbox-icon {
  display: none;
  color: var(--checkbox-icon-color, white);
  width: 0.75rem;
  height: 0.75rem;
}

.smolitux-checkbox-input:checked + .smolitux-checkbox-control .smolitux-checkbox-icon--checked,
.smolitux-checkbox-input:indeterminate + .smolitux-checkbox-control .smolitux-checkbox-icon--indeterminate {
  display: block;
}

.smolitux-checkbox-label {
  margin-left: 0.5rem;
  font-size: 0.875rem;
  color: var(--checkbox-label-color, #374151);
  user-select: none;
}

.smolitux-checkbox-container--label-left {
  flex-direction: row-reverse;
}

.smolitux-checkbox-container--label-left .smolitux-checkbox-label {
  margin-left: 0;
  margin-right: 0.5rem;
}

.smolitux-checkbox-helper-text {
  font-size: 0.75rem;
  color: var(--checkbox-helper-text-color, #6b7280);
  margin-top: 0.25rem;
  margin-left: 1.5rem;
}

.smolitux-checkbox-error {
  font-size: 0.75rem;
  color: var(--checkbox-error-color, #ef4444);
  margin-top: 0.25rem;
  margin-left: 1.5rem;
}

.smolitux-checkbox-success-message {
  font-size: 0.75rem;
  color: var(--checkbox-success-color, #10b981);
  margin-top: 0.25rem;
  margin-left: 1.5rem;
}

/* Größen */
.smolitux-checkbox-control--xs {
  width: 0.75rem;
  height: 0.75rem;
}

.smolitux-checkbox-control--xs .smolitux-checkbox-icon {
  width: 0.5rem;
  height: 0.5rem;
}

.smolitux-checkbox-container--xs .smolitux-checkbox-label {
  font-size: 0.75rem;
}

.smolitux-checkbox-control--sm {
  width: 0.875rem;
  height: 0.875rem;
}

.smolitux-checkbox-control--sm .smolitux-checkbox-icon {
  width: 0.625rem;
  height: 0.625rem;
}

.smolitux-checkbox-container--sm .smolitux-checkbox-label {
  font-size: 0.8125rem;
}

.smolitux-checkbox-control--lg {
  width: 1.25rem;
  height: 1.25rem;
  border-width: 2px;
}

.smolitux-checkbox-control--lg .smolitux-checkbox-icon {
  width: 0.875rem;
  height: 0.875rem;
}

.smolitux-checkbox-container--lg .smolitux-checkbox-label {
  font-size: 1rem;
}

.smolitux-checkbox-control--xl {
  width: 1.5rem;
  height: 1.5rem;
  border-width: 2px;
}

.smolitux-checkbox-control--xl .smolitux-checkbox-icon {
  width: 1rem;
  height: 1rem;
}

.smolitux-checkbox-container--xl .smolitux-checkbox-label {
  font-size: 1.125rem;
}

/* Varianten */
.smolitux-checkbox-control--outline {
  background-color: transparent;
}

.smolitux-checkbox-input:checked + .smolitux-checkbox-control--outline {
  background-color: transparent;
  border-color: var(--checkbox-checked-border-color, #3b82f6);
}

.smolitux-checkbox-input:checked + .smolitux-checkbox-control--outline .smolitux-checkbox-icon {
  color: var(--checkbox-checked-border-color, #3b82f6);
}

.smolitux-checkbox-control--filled {
  background-color: var(--checkbox-filled-bg, #f3f4f6);
  border-color: transparent;
}

.smolitux-checkbox-control--minimal {
  border-color: transparent;
  background-color: transparent;
}

.smolitux-checkbox-input:checked + .smolitux-checkbox-control--minimal {
  background-color: transparent;
  border-color: transparent;
}

.smolitux-checkbox-input:checked + .smolitux-checkbox-control--minimal .smolitux-checkbox-icon {
  color: var(--checkbox-checked-border-color, #3b82f6);
}

/* Farbschemata */
.smolitux-checkbox-input:checked + .smolitux-checkbox-control--primary,
.smolitux-checkbox-input:indeterminate + .smolitux-checkbox-control--primary {
  background-color: var(--primary-500, #3b82f6);
  border-color: var(--primary-500, #3b82f6);
}

.smolitux-checkbox-input:checked + .smolitux-checkbox-control--secondary,
.smolitux-checkbox-input:indeterminate + .smolitux-checkbox-control--secondary {
  background-color: var(--secondary-500, #8b5cf6);
  border-color: var(--secondary-500, #8b5cf6);
}

.smolitux-checkbox-input:checked + .smolitux-checkbox-control--success,
.smolitux-checkbox-input:indeterminate + .smolitux-checkbox-control--success {
  background-color: var(--green-500, #10b981);
  border-color: var(--green-500, #10b981);
}

.smolitux-checkbox-input:checked + .smolitux-checkbox-control--danger,
.smolitux-checkbox-input:indeterminate + .smolitux-checkbox-control--danger {
  background-color: var(--red-500, #ef4444);
  border-color: var(--red-500, #ef4444);
}

.smolitux-checkbox-input:checked + .smolitux-checkbox-control--warning,
.smolitux-checkbox-input:indeterminate + .smolitux-checkbox-control--warning {
  background-color: var(--yellow-500, #f59e0b);
  border-color: var(--yellow-500, #f59e0b);
}

.smolitux-checkbox-input:checked + .smolitux-checkbox-control--info,
.smolitux-checkbox-input:indeterminate + .smolitux-checkbox-control--info {
  background-color: var(--blue-500, #3b82f6);
  border-color: var(--blue-500, #3b82f6);
}

/* Anzeigetypen */
.smolitux-checkbox-control--switch {
  width: 2.5rem;
  height: 1.25rem;
  border-radius: 9999px;
  justify-content: flex-start;
  padding: 0.125rem;
}

.smolitux-checkbox-control--switch::before {
  content: '';
  position: absolute;
  top: 0.125rem;
  left: 0.125rem;
  width: 1rem;
  height: 1rem;
  border-radius: 50%;
  background-color: var(--checkbox-switch-handle-color, #d1d5db);
  transition: transform 0.2s ease-in-out, background-color 0.2s ease-in-out;
}

.smolitux-checkbox-input:checked + .smolitux-checkbox-control--switch::before {
  transform: translateX(1.25rem);
  background-color: white;
}

.smolitux-checkbox-control--radio {
  border-radius: 50%;
}

.smolitux-checkbox-control--radio::before {
  content: '';
  width: 0;
  height: 0;
  border-radius: 50%;
  background-color: white;
  transition: all 0.2s ease-in-out;
}

.smolitux-checkbox-input:checked + .smolitux-checkbox-control--radio::before {
  width: 0.5rem;
  height: 0.5rem;
}

.smolitux-checkbox-control--toggle {
  width: 3rem;
  height: 1.5rem;
  border-radius: 9999px;
  justify-content: flex-start;
  padding: 0.25rem;
  background-color: var(--checkbox-toggle-bg, #e5e7eb);
}

.smolitux-checkbox-control--toggle::before {
  content: '';
  position: absolute;
  top: 0.25rem;
  left: 0.25rem;
  width: 1rem;
  height: 1rem;
  border-radius: 50%;
  background-color: white;
  box-shadow: 0 1px 2px 0 rgba(0, 0, 0, 0.05);
  transition: transform 0.2s ease-in-out;
}

.smolitux-checkbox-input:checked + .smolitux-checkbox-control--toggle::before {
  transform: translateX(1.5rem);
}

.smolitux-checkbox-control--button {
  width: auto;
  height: auto;
  padding: 0.5rem 1rem;
  border-radius: 0.375rem;
  background-color: var(--checkbox-button-bg, #f3f4f6);
  border-color: var(--checkbox-button-border-color, #d1d5db);
}

.smolitux-checkbox-input:checked + .smolitux-checkbox-control--button {
  background-color: var(--checkbox-button-checked-bg, #3b82f6);
  border-color: var(--checkbox-button-checked-border-color, #3b82f6);
  color: white;
}

.smolitux-checkbox-control--button .smolitux-checkbox-icon {
  display: none;
}

/* Zustände */
.smolitux-checkbox-container--invalid .smolitux-checkbox-control {
  border-color: var(--checkbox-error-border-color, #ef4444);
}

.smolitux-checkbox-container--valid .smolitux-checkbox-control {
  border-color: var(--checkbox-success-border-color, #10b981);
}

.smolitux-checkbox-container--success .smolitux-checkbox-control {
  border-color: var(--checkbox-success-border-color, #10b981);
}

.smolitux-checkbox-container--loading {
  pointer-events: none;
}

.smolitux-checkbox-container--loading::after {
  content: '';
  position: absolute;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  background-color: rgba(255, 255, 255, 0.5);
  border-radius: 0.25rem;
}

.smolitux-checkbox-container--loading::before {
  content: '';
  position: absolute;
  top: 50%;
  left: 50%;
  transform: translate(-50%, -50%);
  width: 1rem;
  height: 1rem;
  border: 2px solid var(--primary-200, #bfdbfe);
  border-top-color: var(--primary-500, #3b82f6);
  border-radius: 50%;
  animation: smolitux-checkbox-spin 1s linear infinite;
  z-index: 1;
}

@keyframes smolitux-checkbox-spin {
  to {
    transform: translate(-50%, -50%) rotate(360deg);
  }
}

/* Hover-Effekt */
.smolitux-checkbox-container--hoverable:hover .smolitux-checkbox-control:not(.smolitux-checkbox-control--disabled) {
  border-color: var(--checkbox-hover-border-color, #9ca3af);
}

/* Dark Mode */
.dark .smolitux-checkbox-control {
  border-color: var(--checkbox-border-color-dark, #4b5563);
  background-color: var(--checkbox-bg-dark, #1f2937);
}

.dark .smolitux-checkbox-label {
  color: var(--checkbox-label-color-dark, #e5e7eb);
}

.dark .smolitux-checkbox-helper-text {
  color: var(--checkbox-helper-text-color-dark, #9ca3af);
}

.dark .smolitux-checkbox-input:disabled + .smolitux-checkbox-control {
  background-color: var(--checkbox-disabled-bg-dark, #374151);
  border-color: var(--checkbox-disabled-border-color-dark, #4b5563);
}

.dark .smolitux-checkbox-control--filled {
  background-color: var(--checkbox-filled-bg-dark, #374151);
}

.dark .smolitux-checkbox-control--switch::before {
  background-color: var(--checkbox-switch-handle-color-dark, #6b7280);
}

.dark .smolitux-checkbox-control--toggle {
  background-color: var(--checkbox-toggle-bg-dark, #4b5563);
}

.dark .smolitux-checkbox-control--button {
  background-color: var(--checkbox-button-bg-dark, #374151);
  border-color: var(--checkbox-button-border-color-dark, #4b5563);
}

.dark .smolitux-checkbox-container--hoverable:hover .smolitux-checkbox-control:not(.smolitux-checkbox-control--disabled) {
  border-color: var(--checkbox-hover-border-color-dark, #6b7280);
}

.dark .smolitux-checkbox-container--loading::after {
  background-color: rgba(31, 41, 55, 0.5);
}