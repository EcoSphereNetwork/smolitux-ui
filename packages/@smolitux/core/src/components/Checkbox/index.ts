export { Checkbox } from "./Checkbox";
export type { 
  CheckboxProps, 
  CheckboxSize, 
  CheckboxVariant, 
  CheckboxColorScheme, 
  CheckboxLabelPosition,
  CheckboxDisplayType
} from "./Checkbox";

// Fur Abwartskompatibilitat mit dem bestehenden Export
export { Checkbox as default } from "./Checkbox";
