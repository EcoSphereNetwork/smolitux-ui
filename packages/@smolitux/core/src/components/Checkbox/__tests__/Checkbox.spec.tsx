import React from 'react';
import renderer from 'react-test-renderer';
import { Checkbox } from '../Checkbox';

describe('Checkbox Snapshots', () => {
  test('renders default checkbox', () => {
    const tree = renderer.create(<Checkbox label="Check me" />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('renders checked checkbox', () => {
    const tree = renderer.create(<Checkbox label="Check me" checked />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
