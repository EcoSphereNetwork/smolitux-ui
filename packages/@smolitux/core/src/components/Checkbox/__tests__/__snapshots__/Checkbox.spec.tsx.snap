// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checkbox Snapshots renders checked checkbox 1`] = `
<div
  className="flex items-start  "
  data-testid="checkbox-container"
  onMouseDown={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  onMouseUp={[Function]}
>
  <div
    className="flex items-center "
  >
    <div
      className="relative"
    >
      <input
        checked={true}
        className="h-5 w-5 bg-white dark:bg-gray-700 text-primary-600 dark:text-primary-500 focus:ring-primary-500 dark:focus:ring-primary-400 rounded hover:border-gray-400 dark:hover:border-gray-500 focus:outline-none focus:ring-2 transition duration-150 ease-in-out cursor-pointer"
        disabled={false}
        id="checkbox-checked"
        onBlur={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        required={false}
        type="checkbox"
      />
    </div>
  </div>
  <div
    className="ml-2"
    data-testid="label-container"
  >
    <div
      className=""
    >
      <label
        className="
            text-base
            font-medium text-gray-700 dark:text-gray-300
            cursor-pointer
            
          "
        data-testid="checkbox-label"
        htmlFor="checkbox-checked"
      >
        Check me
      </label>
    </div>
  </div>
</div>
`;

exports[`Checkbox Snapshots renders default checkbox 1`] = `
<div
  className="flex items-start  "
  data-testid="checkbox-container"
  onMouseDown={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  onMouseUp={[Function]}
>
  <div
    className="flex items-center "
  >
    <div
      className="relative"
    >
      <input
        className="h-5 w-5 bg-white dark:bg-gray-700 text-primary-600 dark:text-primary-500 focus:ring-primary-500 dark:focus:ring-primary-400 rounded hover:border-gray-400 dark:hover:border-gray-500 focus:outline-none focus:ring-2 transition duration-150 ease-in-out cursor-pointer"
        disabled={false}
        id="checkbox-default"
        onBlur={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        required={false}
        type="checkbox"
      />
    </div>
  </div>
  <div
    className="ml-2"
    data-testid="label-container"
  >
    <div
      className=""
    >
      <label
        className="
            text-base
            font-medium text-gray-700 dark:text-gray-300
            cursor-pointer
            
          "
        data-testid="checkbox-label"
        htmlFor="checkbox-default"
      >
        Check me
      </label>
    </div>
  </div>
</div>
`;
